File: java/src/P034/SearchforaRange/SearchforaRange.java
package P034.SearchforaRange;

public class SearchforaRange {
    public int[] searchRange(int[] nums, int target) {
        int start = firstGTE(nums, target);
        if (start == nums.length || nums[start] != target) {
            return new int[] {-1, -1};
        }
        return new int[] {start, firstGTE(nums, target + 1) - 1};
    }

    private static int firstGTE(int[] nums, int target) {
        int lo = 0, hi = nums.length;
        while (lo < hi) {
            int mid = lo + ((hi - lo) >>> 1);
            if (nums[mid] < target) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }
}

File: python/P034_SearchforaRange/SearchforaRange.py
# -*- coding: utf-8 -*-

class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """

        def first_gte(nums, target):
            lo, hi = 0, len(nums)
            while lo < hi:
                mid = (lo + hi) / 2
                if nums[mid] < target:
                    lo = mid + 1
                else:
                    hi = mid
            return lo

        start = first_gte(nums, target)
        if start == len(nums) or nums[start] != target:
            return [-1, -1]
        else:
            return [start, first_gte(nums, target + 1) - 1]